reps/2023-02-03-node-affinity-feature-enhancements.md

```diff
@@ -0,0 +1,399 @@
+And users often write wrong sql when using it.
+```python
+# NodeAffinity use case
+actor_1 = Actor.options(
+        scheduling_strategy=NodeAffinity("gpu_type in [A100, T100]")
+    ).remote()
+```
+
+### Example
+
+**1. Affinity to the node of the specified Instance.**  
+```
+# m6in.2xlarge 8C32G instance node
+ray start ... --labels={"instance": "m6in.2xlarge"}
+
+# m6in.large 2C8G instance node
+ray start ... --labels={"instance": "m6in.large"}
+
+actor_1 = Actor.options(
+        scheduling_strategy=node_affinity(label_in(key="instance", values=["m6in.large"], is_soft=false))
+    ).remote()
+
+```
+
+**2. Anti-Affinity to the nodes of the special node_id.**  
+```
+actor_1 = Actor.options(
+        scheduling_strategy=node_affinity(label_not_in(key="node_id", values=["xxxxx", "aaaaa"], is_soft=false))
+    ).remote()
+```
+
+**3. Anti-Affinity to the node which have GPU.**  
+```
+actor_1 = Actor.options(
+        scheduling_strategy=node_affinity(label_does_not_exist(key="gpu_type", is_soft=false))
+    ).remote()
+```
+
+**4. It must be scheduled to a node whose region is china, and as much as possible to a node whose availability zone is china-east-1.**  
+
+```
+actor_1 = Actor.options(
+        scheduling_strategy=node_affinity([
+            label_in("region", ["china"], is_soft=false),
+            label_in("zone", ["china-east-1"], is_soft=true)
+        ])
+    ).remote()
+```
+**5. NodeAffinity can be used together with the CustomResource mechanism.**  
+These two mechanisms are completely non-conflicting.  
+```
+actor_1 = Actor.options(
+                num_gpus=1,
+                resources={"custom_resource": 1},
+                scheduling_strategy=node_affinity([
+                    label_exist("gpu_type", is_soft=false),
+                    label_in("gpu_type", ["A100", "P100"], is_soft=true),
+                ]).remote()
+```
+
+**FAQ:**  
+This node static labels seems to be very similar to custom_resources. Why not use custom_resources to instance of node_affinity?
+
+1. The semantics of static labels and custom_resources are completely different. custom_resources is a countable resource for nodes. And labels is an attribute or identifier of a node.
+2. Now custom_resources cannot implement the function of anti-affinity.
+3. Using labels to classify nodes and implement affinity/anti-affinity will be more natural and easier for users to accept.
+
+### Implementation plan
+
+![NodeAffinity Concept](https://user-images.githubusercontent.com/11072802/217133358-19ddc916-bf15-4f69-96e1-9cbd157b0e1d.png)
+
+1. Add the labels to node info
+```
+message GcsNodeInfo {
+  // The ID of node.
+  bytes node_id = 1;
+  ...
+  // The user-provided identifier or name for this node.
+  string node_name = 12;
+  // The static labels of nodes.
+  map<String, String> labels = 13;
+}
+```
+2. Add the labels data structure to the resource synchronization data structure(NodeResources). 
+```
+NodeResources {
```

Comment:
Wouldn't we need to add the label shapes requested by tasks to the ResourceRequests?

Reply:
This is for synchronizing the Node Labels information to each raylet node in the Raylet scheduling mode.
1. Putting it in NodeResources can reuse the existing resource synchronization mechanism without creating a new label synchronization mechanism
2. Adding labels here can also allow ActorAffinity to reuse this mechanism.
3. Under the Raylet scheduling mechanism, after the scheduling is completed, the local node needs to confirm whether the final resources meet the requirements. So it is convenient to put it here for final confirmation.

Reply:
Since ResourceRequest is also used for autoscaling, seemingly we need to add label requests as well.

Reply:
Ok I'll try to update a rough design by next week